src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -436,6 +437,16 @@ private async Task StopAsyncCore(bool disposing)
             }
         }
 
+#if NETCOREAPP3_0
+        public async IAsyncEnumerable<object> StreamAsyncCore(string methodName, Type returnType, object[] args, CancellationToken cancellationToken = default)
+        {
+            var stream = (await StreamAsChannelCoreAsyncCore(methodName, returnType, args, cancellationToken)).ReadAllAsync();
+            await foreach(var item in AsyncEnumerableAdapters.MakeCancelableAsyncEnumerable(stream, cancellationToken)) {
```

Comment:
Is there a better way to do this?

Reply:
Instead of iterating over MakeCancelableAsyncEnumerable()'s return value, instead just returning it directly.

You might want to create another version of the method which [actually uses the CancellationToken passed to GetAsyncEnumerator()](https://github.com/aspnet/AspNetCore/blob/1cf98a5b5c533d61d4b5c3681680157733dbb244/src/SignalR/server/Core/src/Internal/AsyncEnumerableAdapters.cs#L39-L41), and returns `IAsyncEnumerable<TResult>` instead of `IAsyncEnumerable<object>`. Returning a `IAsyncEnumerable<TResult>` is actually easier since you can always directly return the value from `ChannelReader<TResult>.ReadAllAsync().GetAsyncEnumerator()` from `CancelableAsyncEnumerable<TResult>.GetAsyncEnumerator()` without worrying about wrapping the async enumerators for value types with BoxedAsyncEnumerator.

Reply:
> Instead of iterating over MakeCancelableAsyncEnumerable()'s return value, just return it directly.

That doesn't really work because of the `await` of `StreamAsChannelCoreAsyncCore` beforehand though, right? That's why this has to be transformed into a generator using the `yield` syntax.

Reply:
If we return the IAsyncEnumerable directly, we'd have to make StreamAsync return `Task<IAsyncEnumerable<TResult>>` which @anures and I decided isn't great. I think using an iterator method is OK then.

We still should use a version of `MakeCancelableAsyncEnumerable<TResult>` that returns `IAsyncEnumerable<TResult>` and uses the CancellationToken passed through to `GetAsyncEnumerator(CancellationToken)` linking it with the CancellationToken from `StreamAsyncCore` if both are non-default.

Reply:
> Task<IAsyncEnumerable<TResult>> which @anurse and I decided isn't great. I think using an iterator method is OK then.

The problem with this is that if we use an iterator and yield return from the result of `MakeCancelableAsyncEnumerable` we actually lose the cancellation token which defeats the purpose.

>We still should use a version of MakeCancelableAsyncEnumerable<TResult> that returns IAsyncEnumerable<TResult>

I worked down this route and realized that the streaming methods like `StreamAsChannelCoreAsync` all return `ChannelReader<object>`. This means that no matter what we would have to iterate and cast on every item.